File: rxbinding-support-v4/src/main/java/com/jakewharton/rxbinding2/support/v4/widget/NestedScrollViewScrollChangeEventObservable.java
package com.jakewharton.rxbinding2.support.v4.widget;

import android.support.v4.widget.NestedScrollView;
import android.support.v4.widget.NestedScrollView.OnScrollChangeListener;
import com.jakewharton.rxbinding2.view.ViewScrollChangeEvent;
import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.android.MainThreadDisposable;

import static io.reactivex.android.MainThreadDisposable.verifyMainThread;

final class NestedScrollViewScrollChangeEventObservable extends Observable<ViewScrollChangeEvent> {
  private final NestedScrollView view;

  NestedScrollViewScrollChangeEventObservable(NestedScrollView view) {
    this.view = view;
  }

  @Override protected void subscribeActual(Observer<? super ViewScrollChangeEvent> observer) {
    verifyMainThread();
    Listener listener = new Listener(view, observer);
    observer.onSubscribe(listener);
    view.setOnScrollChangeListener(listener);
  }

  static final class Listener extends MainThreadDisposable implements OnScrollChangeListener {
    private final NestedScrollView view;
    private final Observer<? super ViewScrollChangeEvent> observer;

    Listener(NestedScrollView view, Observer<? super ViewScrollChangeEvent> observer) {
      this.view = view;
      this.observer = observer;
    }

    @Override
    public void onScrollChange(NestedScrollView v, int scrollX, int scrollY, int oldScrollX,
        int oldScrollY) {
      if (!isDisposed()) {
        observer.onNext(
            ViewScrollChangeEvent.create(view, scrollX, scrollY, oldScrollX, oldScrollY));
      }
    }

    @Override protected void onDispose() {
      view.setOnScrollChangeListener((OnScrollChangeListener) null);
    }
  }
}

File: rxbinding-core/src/main/java/com/jakewharton/rxbinding3/core/NestedScrollViewScrollChangeEventObservable.kt
@file:JvmName("RxNestedScrollView")
@file:JvmMultifileClass

package com.jakewharton.rxbinding3.core

import androidx.annotation.CheckResult
import androidx.core.widget.NestedScrollView
import androidx.core.widget.NestedScrollView.OnScrollChangeListener
import com.jakewharton.rxbinding3.view.ViewScrollChangeEvent
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.Observer
import io.reactivex.rxjava3.android.MainThreadDisposable

import com.jakewharton.rxbinding3.internal.checkMainThread

/**
 * Create an observable of scroll-change events for `view`.
 *
 * *Warning:* The created observable keeps a strong reference to `view`.
 * Unsubscribe to free this reference.
 */
@CheckResult
fun NestedScrollView.scrollChangeEvents(): Observable<ViewScrollChangeEvent> {
  return NestedScrollViewScrollChangeEventObservable(this)
}

private class NestedScrollViewScrollChangeEventObservable(
  private val view: NestedScrollView
) : Observable<ViewScrollChangeEvent>() {

  override fun subscribeActual(observer: Observer<in ViewScrollChangeEvent>) {
    if (!checkMainThread(observer)) {
      return
    }
    val listener = Listener(view, observer)
    observer.onSubscribe(listener)
    view.setOnScrollChangeListener(listener)
  }

  private class Listener(
    private val view: NestedScrollView,
    private val observer: Observer<in ViewScrollChangeEvent>
  ) : MainThreadDisposable(), OnScrollChangeListener {

    override fun onScrollChange(
      v: NestedScrollView, scrollX: Int, scrollY: Int, oldScrollX: Int,
      oldScrollY: Int
    ) {
      if (!isDisposed) {
        observer.onNext(ViewScrollChangeEvent(view, scrollX, scrollY, oldScrollX, oldScrollY))
      }
    }

    override fun onDispose() {
      view.setOnScrollChangeListener(null as OnScrollChangeListener?)
    }
  }
}
